src/client/mod.rs

```diff
@@ -780,41 +901,55 @@ impl IotHubClient {
                     }
                 }
-
-                Result::Err(e) => error = e.to_string(),
-            }
-        } else {
-            error = format!("method not implemented: {}", method_name)
-        }
-
-        error!("{}", error);
-
-        match CString::new(json!(error).to_string()) {
-            Ok(r) => {
-                *response_size = r.as_bytes().len();
-                *response = r.into_raw() as *mut u8;
-            }
-            Err(e) => {
-                error!("cannot parse direct method result: {}", e);
+                Ok(Err(e)) => match CString::new(json!(e.to_string()).to_string()) {
+                    Ok(r) => {
+                        *response_size = r.as_bytes().len();
+                        *response = r.into_raw() as *mut u8;
+                    }
+                    Err(e) => {
+                        error!("cannot parse direct method result: {e}");
+                    }
+                },
+                Err(e) => {
+                    error!("channel unexpectedly closed: {e}");
+                }
             }
         }
 
         METHOD_RESPONSE_ERROR
     }
 
     unsafe extern "C" fn c_d2c_confirmation_callback(
-        status: IOTHUB_CLIENT_RESULT,
-        ctx: *mut std::ffi::c_void,
+        status: IOTHUB_CLIENT_CONFIRMATION_RESULT,
+        context: *mut std::ffi::c_void,
     ) {
-        debug!(
-            "Received confirmation from iothub for event with internal id: {} and status: {}",
-            ctx as u32, status
-        );
+        let result: Box<(Arc<Mutex<bool>>, Arc<Notify>)> =
+            Box::from_raw(context as *mut (Arc<Mutex<bool>>, Arc<Notify>));
+
+        match status {
+            IOTHUB_CLIENT_CONFIRMATION_RESULT_TAG_IOTHUB_CLIENT_CONFIRMATION_OK => {
+                *result
+                    .0
+                    .lock()
+                    .expect("c_d2c_confirmation_callback: cannot lock result") = true;
+                debug!("c_d2c_confirmation_callback: received confirmation from iothub.");
+            },
+            IOTHUB_CLIENT_CONFIRMATION_RESULT_TAG_IOTHUB_CLIENT_CONFIRMATION_BECAUSE_DESTROY => error!("c_d2c_confirmation_callback: received confirmation from iothub with error IOTHUB_CLIENT_CONFIRMATION_BECAUSE_DESTROY."),
+            IOTHUB_CLIENT_CONFIRMATION_RESULT_TAG_IOTHUB_CLIENT_CONFIRMATION_ERROR =>  error!("c_d2c_confirmation_callback: received confirmation from iothub with error IOTHUB_CLIENT_CONFIRMATION_ERROR."),
+            IOTHUB_CLIENT_CONFIRMATION_RESULT_TAG_IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT => error!("c_d2c_confirmation_callback: received confirmation from iothub with error IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT."),
+            _ => {
+                error!("c_d2c_confirmation_callback: received confirmation from iothub with unknown IOTHUB_CLIENT_CONFIRMATION_RESULT");
+                return;
+            }
+        }
+
+        result.1.notify_one();
+
+        drop(result);
```

Comment:
Per https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw the destructor of this box should be called automatically, why is the extra drop here necessary?

Reply:
long story but you seem to be right... anyway switched back to oneshot channel